aspnet/mvc/models/validation.rst

```diff
@@ -1,8 +1,196 @@
-============================
+In this article:
 
-.. include:: /../common/stub-notice.txt
+.. contents:: Sections
+  :local:
+  :depth: 1
 
-.. _issue: https://github.com/aspnet/Docs/issues/116
+Introduction to model validation
+--------------------------------
+Before an app stores data in a database, the app must validate the data. Data must be scrubbed for security, verified that it is appropriately formatted by type and size, and it must conform to your rules. Validation is necessary although it can be redundant and tedious to implement. In MVC, validation happens on both the client and server. 
+
+With data annotations in MVC apps you can validate data by customizing models so they can enforce business rules against models by applying attributes to model properties. These attributes generate validation code, thereby reducing the amount of code you must write. Fortunately, .NET has abstracted validation into the data annotations, often reducing validation to a single line of code.
+
+Data annotations
+----------------
+Data annotations are a way to configure model validation so it's similar conceptually to validation on fields in tables. This includes constraints such as assigning data types or required fields. Other types of validation include applying patterns to data to enforce business rules, such as a credit card, phone number, or email address (regular expressions). Data annotations do all this for you automatically.
+
+Below is an annotated ``Movie`` model from an app that stores information about movies and TV shows. Most of the properties are required, the ``Id`` property is a primary key, and several string properties have length requirements. Additionally, there is a numeric range in place for the ``Price`` property from 0 to $999.99, along with a custom annotation.
+
+.. literalinclude:: validation/sample/Movie.cs
+   :language: c#
+   :lines: 7-34
+   :dedent: 4
+
+Simply reading through the model reveals the rules about data for this app, making it easier to maintain the code. Below are several popular built-in data annotations:
+
+- ``[CreditCard]``: Formats the property as a credit card type.
+- ``[Compare]``: Compares two properties in a model. It works with various types, such as int, string, or date. 
+- ``[Email]``: Formats the property with an email format.
+- ``[Phone]``: Formats the property as a telephone format.
+- ``[Range]``: Sets the minimum and maximum values for multiple types of properties.
+- ``[RegularExpression]``: Forces the data to match a regular expression.
+- ``[Required]``: Makes a property required.
+- ``[StringLength]``: The maximum length for a string field.
+- ``[Url]``: Formats the property as a URL.
+
+A complete list of annotations is at the `System.ComponentModel.DataAnnotations <https://msdn.microsoft.com/en-us/library/system.componentmodel.dataannotations(v=vs.110).aspx>`_ namespace docs. Note that this namespace is not specific to MVC, but in a shared .NET namespace. 
```

Comment:
@danroth27 the MVC validation system supports all `[Validation]`s applied to model types and properties server-side.

Restrictions come in on the client side: MVC supports _most_ `[Validation]`s when generating the requisite HTML attributes. The full list of supported `[Validation]`s client-side excludes mainly `[CustomValidation]`s and any (custom obviously) subclass. MVC also lacks support for the `[DataType]` (the base attribute), `[EnumDataType]` and `[FileExtensions]`.

Might be worth reminding users that a base `[DataType]` may cause client-side validation to occur because e.g. `<input type="phone" ..>` changes requirements browsers enforce. But this attribute does _not_ cause server-side validation because its `IsValid()` method does almost nothing -- all the smarts are in subclasses like `[Phone]`. That is MVC supports this attribute but it doesn't do much in the server-side validation system.